data/src/test/java/org/apache/iceberg/data/avro/TestReadDefaultValues.java

```diff
@@ -0,0 +1,198 @@
+        // {Types.DecimalType.of(9, -20), "\"2E+20\""},
+        {Types.ListType.ofOptional(1, Types.IntegerType.get()), "[1, 2, 3]"},
+        {
+          Types.MapType.ofOptional(2, 3, Types.IntegerType.get(), Types.StringType.get()),
+          "{\"keys\": [1, 2], \"values\": [\"foo\", \"bar\"]}"
+        },
+        {
+          Types.StructType.of(
+              required(4, "f1", Types.IntegerType.get()),
+              optional(5, "f2", Types.StringType.get())),
+          "{\"4\": 1, \"5\": \"bar\"}"
+        },
+        // deeply nested complex types
+        {
+          Types.ListType.ofOptional(
+              6,
+              Types.StructType.of(
+                  required(7, "f1", Types.IntegerType.get()),
+                  optional(8, "f2", Types.StringType.get()))),
+          "[{\"7\": 1, \"8\": \"bar\"}, {\"7\": 2, \"8\": " + "\"foo\"}]"
+        },
+        {
+          Types.MapType.ofOptional(
+              9,
+              10,
+              Types.IntegerType.get(),
+              Types.StructType.of(
+                  required(11, "f1", Types.IntegerType.get()),
+                  optional(12, "f2", Types.StringType.get()))),
+          "{\"keys\": [1, 2], \"values\": [{\"11\": 1, \"12\": \"bar\"}, {\"11\": 2, \"12\": \"foo\"}]}"
+        },
+        {
+          Types.StructType.of(
+              required(
+                  13,
+                  "f1",
+                  Types.StructType.of(
+                      optional(14, "ff1", Types.IntegerType.get()),
+                      optional(15, "ff2", Types.StringType.get()))),
+              optional(
+                  16,
+                  "f2",
+                  Types.StructType.of(
+                      optional(17, "ff1", Types.StringType.get()),
+                      optional(18, "ff2", Types.IntegerType.get())))),
+          "{\"13\": {\"14\": 1, \"15\": \"bar\"}, \"16\": {\"17\": \"bar\", \"18\": 1}}"
+        },
+      };
+
+  @Test
+  public void writeAndValidate() throws IOException {
+    for (Object[] typeWithDefault : typesWithDefaults) {
+      Type type = (Type) typeWithDefault[0];
+      String defaultValueJson = (String) typeWithDefault[1];
+      Object defaultValue = SingleValueParser.fromJson(type, defaultValueJson);
+
+      Schema writerSchema = new Schema(required(999, "col1", Types.IntegerType.get()));
+
+      Schema readerSchema =
+          new Schema(
+              required(999, "col1", Types.IntegerType.get()),
+              NestedFieldWithInitialDefault.optionalWithDefault(
+                  1000, "col2", type, null, defaultValue, defaultValue));
+
+      List<Record> generatedRecords = RandomGenericData.generate(writerSchema, 100, 0L);
+      List<Record> expected = Lists.newArrayList();
+      for (Record record : generatedRecords) {
+        Record expectedRecord = GenericRecord.create(readerSchema);
+        expectedRecord.set(0, record.get(0));
+        expectedRecord.set(1, defaultValue);
```

Comment:
I think the problem with fixed might be introduced here. Iceberg generics will not produce the same value as the internal representation. If you expect `ByteBuffer` here, the generics are going to produce `ByteBuffer`.

You could stop setting the reader func to `DataReader::create` below, but then the `GenericRecord` instances would not be produced. What I'd do here is have a `convert` method to convert in this test all of the types that don't match between the two, like timestamps, dates, and fixed.

Reply:
Addressed per above discussion.